drivers/modem/murata-1sc.c

```diff
@@ -0,0 +1,4499 @@
+}
+
+/**
+ * @brief Convert uint8_t to a binary ascii (characters of either 1 or 0).
+ */
+static size_t hex_str_to_data(const char *input_buf, uint8_t *output_buf, size_t output_len)
+{
+	size_t str_len = strlen(input_buf);
+	size_t i = 0;
+
+	for (i = 0; (i < output_len) && (i * 2 < str_len); i++) {
+		output_buf[i] = hex_byte_to_data(&input_buf[i * 2]);
+	}
+	return i;
+}
+
+/**
+ * @brief Convert uint8_t to a binary ascii (string with 1 and 0).
+ */
+static char *byte_to_binary_str(uint8_t byte)
+{
+	static char buf[9] = {0};
+
+	memset(buf, 0, sizeof(buf));
+	for (int i = 0; i < 8; i++) {
+		buf[7 - i] = (byte & 1 << i) ? '1' : '0';
+	}
+
+	return buf;
+}
+
+/* Modem pins - Wake Host, Wake Modem, Reset, and Reset Done */
+static const struct gpio_dt_spec wake_host_gpio =
+	GPIO_DT_SPEC_INST_GET_OR(0, mdm_wake_host_gpios, {0});
+static const struct gpio_dt_spec wake_mdm_gpio =
+	GPIO_DT_SPEC_INST_GET_OR(0, mdm_wake_mdm_gpios, {0});
+static const struct gpio_dt_spec reset_gpio = GPIO_DT_SPEC_INST_GET_OR(0, mdm_reset_gpios, {0});
+static const struct gpio_dt_spec rst_done_gpio =
+	GPIO_DT_SPEC_INST_GET_OR(0, mdm_rst_done_gpios, {0});
+static const struct gpio_dt_spec mdm_rx_gpio = GPIO_DT_SPEC_INST_GET_OR(0, mdm_rx_gpios, {0});
+static const struct gpio_dt_spec mdm_tx_gpio = GPIO_DT_SPEC_INST_GET_OR(0, mdm_tx_gpios, {0});
+
+static struct k_thread modem_rx_thread;
+
+#if !defined(CONFIG_MODEM_MURATA_USE_SYSTEM_WORKQUEUE)
+static struct k_work_q modem_workq;
+#endif
+
+static struct murata_1sc_data mdata;
+static struct modem_context mctx;
+static const struct socket_op_vtable offload_socket_fd_op_vtable;
+
+static void socket_close(struct modem_socket *sock);
+#ifdef HIFC_AVAILABLE
+static const char s_hifc_mode = MAX_HIFC_SUPPORTED;
+static const char s_max_pm_mode[] = MAX_PM_MODE;
+#endif
+
+static const bool s_sleep_mode = true;
+
+/* RX thread structures */
+static K_KERNEL_STACK_DEFINE(modem_rx_stack, CONFIG_MODEM_MURATA_1SC_RX_STACK_SIZE);
+NET_BUF_POOL_DEFINE(mdm_recv_pool, MDM_RECV_MAX_BUF, MDM_RECV_BUF_SIZE, 0, NULL);
+
+#if !defined(CONFIG_MODEM_MURATA_USE_SYSTEM_WORKQUEUE)
+static K_KERNEL_STACK_DEFINE(modem_workq_stack, CONFIG_MODEM_MURATA_WORKQ_STACK_SIZE);
+#endif
+
+/**
+ * @brief Thread to process all messages received from the Modem.
+ */
+static void murata_1sc_rx(void)
+{
+	while (true) {
+		/* Wait for incoming data */
+		modem_iface_uart_rx_wait(&mctx.iface, K_FOREVER);
+
+		modem_cmd_handler_process(&mctx.cmd_handler, &mctx.iface);
+	}
+}
+
+/**
+ * @brief Convert string to long integer, but handle errors
+ */
+static int murata_1sc_atoi(const char *s, const int err_value, const char *desc, const char *func)
+{
+	int ret;
+	char *endptr;
+
+	ret = (int)strtol(s, &endptr, 10);
+	if (!endptr || *endptr != '\0') {
+		LOG_ERR("bad %s '%s' in %s", s, desc, func);
+		return err_value;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Convert ascii hex to uint8_t
+ */
+static inline uint8_t hex_char_to_int(char ch)
```

Comment:
looks very identical to [char2hex](https://github.com/zephyrproject-rtos/zephyr/blob/main/lib/os/hex.c#L12)